Flexible string-based data addressing with URL-like "dataspec"

c42f opened the pull request:
This parses the argument to dataset() with a URL-like scheme, enabling
more flexible naming of resources within a given dataset. Like URLs, we
use the & and # as separators for query and fragment sections.

Like URLs, this provides for a compact string representation which can
be easily communicated between users of a shared dataset to address
subresources without the need to create an entirely new and largely
duplicated dataset configuration.

As in URLs, the query section is to be used during resource resolution,
for example a version number for the dataset can be specified.  We
choose the standard & and = separators for key value pairs:

    name/of/dataset?version=v1

As in URLs, the fragment section refers to a subresource and its
interpretation depends on the type of the dataset. (In analogy, the
meaning of the fragment of a URL depends on the mime type received from
the server.) For example, a subtree of a BlobTree can be addressed using

    tree_name#path/to/subtree

The use of fragment for indexing a BlobTree is implemented here as one
handy use case of this syntax.

Closes #32

pfitzseb commented:
So should we just commit to making dataset URI-addressed, e.g. `dataset://juliahub.com/pfitzseb/some_dataset#foo/bar`? The authority and scheme could be optional in some contexts, which would exactly recover what you've implemented here.

c42f commented:
You got it :-)  By reusing URI syntax we get some natural syntax extensions for free and we also get familiarity for anyone familiar with URIs.

Having explicit authorities is interesting and could be a good extension (/ alternative?) to the `DataProject` - based system we have right now.

At the moment, `DataSets.PROJECT` is a stack of data projects, and datasets are searched by name starting at the top of the stack. Projects on the top of the stack can shadow dataset names further down the stack.

But an explicit authority might do something like name a particular `DataProject` within the stack? Or maybe there would be a builtin resolver for `data://` URLs with a well defined protocol for contacting "data servers" in a similar way that we have a Pkg protocol? All just speculation at this point.

With regard to the technical part of this PR, I'm not sure that adding the fragment/query to the `DataSet` configuration is quite an ideal way to do this though it is quite simple. As an alternative, we could return a `DataSpec` object which wraps the `DataSet` and this would avoid clobbering any `dataset.parameters` which already exist.

zot commented on src/DataSets.jl:
```diff
@@ -188,21 +201,86 @@ To open a directory as a browsable tree object,
+    # name/of/dataset?param1=value1&param2=value2#fragment
+    m = match(r"
+        ^
+        ((?:[[:alpha:]][[:alnum:]_]*/?)+)  # name     - a/b/c
```
Do you want to allow leading slashes?

c42f replied:
I don't think so, at least not at the moment.

I think `DataSet` names seem more like relative paths, relative to the data project which contains them. If we had data URLs with an authority section, they'd be relative to that authority.

c42f commented on src/DataSets.jl:
```diff
@@ -188,21 +201,86 @@ To open a directory as a browsable tree object,
+        if c == '%'
+            c1 = read(io, Char)
+            c = read(io, Char)
+            write(out, parse(UInt8, string(c1, c); base=16))
```
True! As the comment says I just stole this function from URIs.jl though! So I can't take real credit for any neatness (or bugs :) ) here.

The other way to do this would be to iterate the input string using `i = nextind(str, i)` and `str[i]`. This is also unicode-safe and avoids the need to create an `IOBuffer` just to read the input.

The other-other way to do it would be to iterate the input string using `Iterators.Stateful(str)`, which can be used statefully, much like the `IOBuffer` is here.

c42f replied:
Oh yes, forget those comments about iterating input — agreed that the output `IOBuffer` is very natural.

c42f commented:
I tweaked a couple of things and added tests.

The dataspec which was previously merged into the `"parameters"` section is now renamed to a `"dataspec"` section. I was imagining that the user might want to pre-specify "parameters" in the config and override them with the dataspec.

But it's not that simple. For example, paths in the fragment `tree#path/to/blob`, should probably concatenate on to the end of any paths existing in the dataset config. But this kind of thing needs some logic, it can't really be done with pure configuration. So we should probably preserve rather than overwrite any existing relative paths present in the config.  This seemed to call for a more specific unambiguous name in the config for parts of the URL and I just called that `dataspec` for now.
